src/bio.c

```diff
@@ -322,3 +406,34 @@ void bioKillThreads(void) {
         }
     }
 }
+
+void bioPipeReadJobCompList(aeEventLoop *el, int fd, void *privdata, int mask) {
+    UNUSED(el);
+    UNUSED(mask);
+    UNUSED(privdata);
+
+    char buf[128];
+    list *tmp_list = NULL;
+
+    while (read(fd, buf, sizeof(buf)) == sizeof(buf));
+
+    /* Handle event loop events if pipe was written from event loop API */
+    pthread_mutex_lock(&bio_mutex_comp);
+    if (listLength(bio_comp_list)) {
+        tmp_list = bio_comp_list;
+        bio_comp_list = listCreate();
+    }
+    pthread_mutex_unlock(&bio_mutex_comp);
+
+    if (!tmp_list) return;
```

Comment:
what about using assert instead?

Reply:
There is a chance that the BG thread will manage to write twice or more to completion list and the main thread will consume them in a single access.

Reply:
not sure i can follow you, `bio_mutex_comp` should have at least one completion job when the main thread was woke up, so `tmp_list` should not be NULL.

Reply:
Consider the following flow:
* Main: bioCreateCompRq(BIO_WORKER_LAZY_FREE, ...);
* Main: bioCreateCompRq(BIO_WORKER_LAZY_FREE, ...);
* BG:  listAddNodeTail(bio_comp_list, ... ) write(job_comp_pipe[1], ...)
* BG:  listAddNodeTail(bio_comp_list, ... ) 
* Main: bioPipeReadJobCompList() // consume both
* BG:  write(job_comp_pipe[1], ...)
* Main: bioPipeReadJobCompList() // consume nothing

We can alternatively add write() into mutex with listAddNodeTail() but maybe it is better not to block too much main thread and just let it do trivial return if no items.

Reply:
yeah, you're right.